### PR TITLE
Feature/225

### PR DESCRIPTION
**Pull request by Jahenstein**
- added login via ldap
- at this moment we store sensitive data about users in db, ldap implementation without such data in db will be introduced in later releases
- contact me before review, I will send you necessary data

### CODE REVIEW

File: views/index.py [hunk 1]
```diff
@@ -61,97 +54,86 @@ def index():
...
+                        user_db.employee_id = user_ldap_data['employeeID']
+                        db.session.commit()
+
+                user_db = User.query.filter_by(
+                    employee_id=user_ldap_data['employeeID']
+                ).first()
+                session['logged_in'] = True
+                session['id'] = user_db.id
+                session['email'] = user_db.email
+                print(session)
```
Comment (wojwarych) [inline views/index.py]: I do not think it's secure to print user's session data to server logs. :)

Comment (Jahenstein) [reply views/index.py]: It is debugging artifact, deleted

File: tests/conftest.py [hunk 2]
```diff
@@ -120,6 +116,94 @@ def email_generator(chars=string.ascii_letters + string.digits + '.' + '-'):
...
+        email = g.person.email(['@tieto.com'])
+        passwd = g.person.password(length=12)
+        name = g.person.name()
+        surname = g.person.last_name()
+        identifier = g.person.identifier(mask='#####')
+
+        if wroclaw_usr:
+            location = 'Wroclaw'
+        else:
+            while True:
```
Comment (pako535) [inline tests/conftest.py]: I do not know if it's not better `while g.address.city() is not 'Wroclaw':`

File: tests/conftest.py [hunk 3]
```diff
@@ -120,6 +116,94 @@ def email_generator(chars=string.ascii_letters + string.digits + '.' + '-'):
...
+                location = g.address.city()
+                if location != 'Wroclaw':
```
Comment (pako535) [inline tests/conftest.py]: Then this if is unnecessary. Now, during the loop, condition you will be checked once not twice

File: tests/conftest.py [hunk 4]
```diff
@@ -120,6 +116,94 @@ def email_generator(chars=string.ascii_letters + string.digits + '.' + '-'):
...
+        def bind_user(self, user_name, passwd):
+            users = [self.user, self.admin, self.user_not_wroc]
+            check_login = [
+                user_name == usr['user_name'][0].decode() and
+                passwd == usr['passwd'][0].decode()
+                for usr in users
+            ]
+            if any(check_login):
+                return True
+            else:
```
Comment (pako535) [inline tests/conftest.py]: delete this else is not necessary. `return False` should be on end of function

File: tests/conftest.py [hunk 5]
```diff
@@ -120,6 +116,94 @@ def email_generator(chars=string.ascii_letters + string.digits + '.' + '-'):
...
+            location = 'Wroclaw'
+        else:
+            while True:
+                location = g.address.city()
+                if location != 'Wroclaw':
+                    break
+
+        if non_user:
+            return dict(user_name=login, passwd=passwd)
+        else:
```
Comment (pako535) [inline tests/conftest.py]: this else is necessary, the code below will not work if 'if' will be true

File: tests/test_book_borrowing_dashboard.py [hunk 6]
```diff
@@ -1,43 +1,54 @@
...
-    assert reserved_items[1]._reservation_begin == \
-        user_reservations[3][1]._reservation_begin
-    assert reserved_items[1]._reservation_end == \
-        user_reservations[3][1]._reservation_end
+def test_get_reserved_items(session, user_reservations, mock_ldap):
+    with mock.patch('views.index.ldap_client', mock_ldap):
+        reserved_items = get_reserved_items(
+            session, user_reservations[0]['db_id']
+        )
+        assert reserved_items[0].LibraryItem == user_reservations[1][0]
```
Comment (pako535) [inline tests/test_book_borrowing_dashboard.py]: You should write your own assertion to get more transparency

File: tests/test_book_borrowing_dashboard.py [hunk 7]
```diff
@@ -1,43 +1,54 @@
...
-    assert borrowed_items[1]._borrow_time == \
-        user_reservations[2][1]._borrow_time
-    assert borrowed_items[1]._return_time == \
-        user_reservations[2][1]._return_time
+def test_get_borrowed_items(session, user_reservations, mock_ldap):
+    with mock.patch('views.index.ldap_client', mock_ldap):
+        borrowed_items = get_borrowed_items(
+            session, user_reservations[0]['db_id']
+        )
+        assert borrowed_items[0].LibraryItem == \
```
Comment (pako535) [inline tests/test_book_borrowing_dashboard.py]: You should write your own assertion to get more transparency. 
IS similar to the above, maybe it's worth writing some clever assertion function

File: tests/test_copy.py [hunk 8]
```diff
@@ -42,30 +48,36 @@ def test_add_post_data_redirect_status_code(
...
+):
+    with mock.patch('views.index.ldap_client', mock_ldap):
+        client.post(url_for('library.login'),
+                    data=login_form_admin_credentials.data)
+        client.post(url_for('library.add_copy',
+                            item_id=db_book.id),
+                    data=copy_form[0].data)
+        copy = Copy.query.filter_by(
+            asset_code=copy_form[0].asset_code.data).first()
+        assert copy, \
```
Comment (pako535) [inline tests/test_copy.py]: A good habit of writing tests is writing your own assertions. Try write own assertion eg. `assert_copies(copy_1, copy_2)`

File: tests/test_copy.py [hunk 9]
```diff
@@ -122,31 +144,38 @@ def test_edit_post_data_redirect_status_code(
...
+        client.post(url_for('library.login'),
+                    data=login_form_admin_credentials.data)
+        client.post(url_for('library.add_copy',
+                            item_id=db_book.id), data=copy_form[0].data)
+        copy_before_edit = Copy.query.get(db_copies[0].id)
+        client.post(url_for('library.edit_copy',
+                            copy_id=db_copies[0].id),
+                    data=copy_form[1].data)
+        copy = Copy.query.get(db_copies[0].id)
+        assert copy, \
```
Comment (pako535) [inline tests/test_copy.py]: The same as above, maybe you can use the same assertion

**Pull request merged**